src/rust/engine/src/externs/mod.rs

```diff
@@ -314,24 +317,99 @@ pub fn call_method(value: &PyObject, method: &str, args: &[Value]) -> Result<PyO
   value.call_method(gil.python(), method, args_tuple, None)
 }
 
-pub fn call_function<T: AsRef<PyObject>>(func: T, args: &[Value]) -> Result<PyObject, PyErr> {
-  let func: &PyObject = func.as_ref();
+///
+/// The result of an intrinsic or @rule is contained in this struct, depending on whether it returns
+/// a value, or raises an exception. This is *not* a Result, because most of the time we want to
+/// pass an exception from one coroutine or functional call right back into a "parent" coroutine,
+/// which allows for try/except handling within an `await Get(...)`.
```

Comment:
It's not clear why that means that it cannot be a `Result<Value, Failure>`. It seems like there is a non-trivial amount of boilerplate involved with this being a unique type.

But also, since it seems like the only type of `Failure` a `@rule` might possibly want to observe would be a `Failure::Throw`, it might be better to convert the `Failure::Throw` enum member into a reusable struct:
```
enum Failure {
  ..
  Throw(Throw)
}

struct Throw {
  // all of the fields of the current Failure::Throw
}
```
... and then have `InvocationResult` be represented as `Result<Value, Throw>`? Unclear... I might be missing something.

Reply:
> It seems like there is a non-trivial amount of boilerplate involved with this being a unique type.

Yes -- that was my intent, to make the transitions more explicit because there are two ways a value can exit a function and it's not always clear which is which.

> .. and then have InvocationResult be represented as Result<Value, Throw>

I like this a lot more. I'll try this.